keps/sig-storage/1845-prioritization-on-volume-capacity/README.md

```diff
@@ -0,0 +1,428 @@
+capacity ratio using the following formula:
+
+```
+[UtilizationPerClass] = [MaxUtilization] * ([RequestPerClass] / [CapacityPerClass])
+```
+
+The utilization ranges from 0 to 100.
+
+We can tune this formula by applying utilization shape points, which is used in
+[`RequestedToCapacityRatio`](https://github.com/kubernetes/kubernetes/blob/master/pkg/scheduler/framework/plugins/noderesources/requested_to_capacity_ratio.go).
+
+By default, we use this shape points to get linear scores between [0,
+100] utilization.
+
+```
+- utilization: 0
+  score: 0
+- utilization: 100
+  score: config.MaxCustomPriorityScore
+```
+
+The score ranges from 0 to `config.MaxCustomPriorityScore`. (Internal score
+will be scaled up to the score range used by the scheduler).
+
+Optionally, users can configure customized utilization shape points, e.g.
+
+```
+- utilization: 50
+  score: 0
+- utilization: 80
+  score: 3
+- utilization: 100
+  score: 5
+```
+
+The score of utilization below 50 will be 0, and the score of utilization of 90
+will be 4 (`3 + (5 - 3) * (90 - 80) / (100 - 80)`).
+
+2) Then, we add them together and divides by the number of storage classes.
+
+### Configuring the utilization shape points
+
+We can add a field `Shape` to specify the shape points which defines the score
+function shape.
+
+```
+// VolumeBindingArgs holds arguments used to configure the VolumeBinding plugin.
+type VolumeBindingArgs struct {
+		...
+    // Shape specifies the shape points defining score function shape
+    Shape []UtilizationShapePoint
+}
+```
+
+The definition of `UtilizationShapePoint` is [here](https://github.com/kubernetes/kubernetes/blob/73fa63a86d1d47d82b435ca85c2f350da06b08b9/pkg/scheduler/apis/config/types_pluginargs.go#L115-L121).
+
+### Configuring the weight of storage class
+
+Optionally, we can allow users to configure the weight of storage class. If
+weights of storage classes are evolved, the formula will be:
+
+```
+[NodeScore] = Sum([ScorePerClass] * [WeightPerClass]) / Sum([WeightPerClass])
+```
+
+In the initial stage, we treat all storage classes equal to avoid introducing
+complexity to Kubernetes. If there are real use cases in Beta/GA, we can
+consider supporting it.
+
+```
+<<[UNRESOLVED]>>
+
+Possible solutions:
+
+1) annotation in storage class object
+
+We can read the weight from the storage class annotation, e.g.
+
+```
+storageclass.kubernetes.io/weight: "5"
+```
+
+Pros:
+
+- No need to update the configuration of the kube-scheduler
+- The change can be applied on the fly
+
+Cons:
+
+- Weights are configured in different objects
+
+2) configure in plugin args:
+
+```
+storageClassWeights:
+- local-ssd: "5"
+- local-hdd: "3"
+```
+
+Pros:
+
+- Weights are configured in one place
+
+Cons:
+
+- Need to update and reload the configuration of the kube-scheduler
+
+<<[/UNRESOLVED]>>
+```
+
+### Test Plan
+
+- **Unit Tests:** All code will be covered by unit tests.
+- **Integration Tests** Typical user cases will be covered in scheduler
```

Comment:
Do you think it would be valuable to also have some benchmarking tests to experiment with different weights and/or scoring functions?

Reply:
I've added a plan to add benchmarking tests for VolumeBinding scenarios with this feature enabled.

the performance of the current algorithm should be fine because the complexity does not change with different weights or shape points. The time is mainly determined by the number of PVCs/PVs involved. I think it's not necessary to experiment with different configurations. But we may need to think about the default value of storage class weight and if there is a better scoring function.